\chapter{Results}\label{chap:results}
\subsubsection{scRNA-seq dataset}
The pre-processed dataset used here contains 3855 genes and 9751 cells. There are 15 different cell types. 
Among these, 14 types are NBs and one is Midline Glia (MG) which is a precursor for Glia cells. 
Fig. \ref{fig:type_count} displays the number of cells per type, arranged in ascending order from the fewest to the most. 
Numbers represent NBs (i.e 3-5 is NB3-5). 

\begin{figure}[ht]
    \centering
    \includegraphics[scale=0.5]{figures/results/type_count.png}
    \caption{Pre-processed scRNA-seq dataset cell types count}
    \label{fig:type_count}
\end{figure}

\section{SCENIC Workflow}
\subsubsection{Inferring a GRN}
As explained in Methodology (section \ref{chap:Methodology}), 
the SCENIC workflow was executed repeatedly 50 times and each run produced a list of predicted regulons, 
and an AUCell cell enrichment score for each individual cell per regulon. 

For each SCENIC iteration, after motif enrichment, 
a list of regulons was saved in a CSV file, with each regulon composed of one TF and its target genes. 
Regulons are named after the TF at their head. 
The occurrences of TFs were aggregated across all runs, identifying 352 unique TFs. 
The histogram in fig. \ref{fig:TFs_freq} shoes how frequently each TF was repeated across different runs. 
The x-axis represents the number of run in which a TF appears, ranging from 1 to 50 (with bin size of five). 
The y-axis indicates the number of TFs that appears in that specific number of runs. 
The red line indicates the threshold set at 80\% of runs (>40 iterations); 
38 TFs repeated in over 40 iterations in regulons and were selected for further analysis, with 3330 unique genes.
It is important to note that between different iterations, regulons with the same TF still differ in target genes composition. 

To check the variability of the genes within the 38  regulons of  the most occurring TFs across iterations, 
each regulon was spread to a list of TF-target couples, and the occurrences of these couples across iterations was counted. 
Similarly to fig, \ref{fig:TFs_freq}, the histogram in fig. \ref{fig:TF-target_freq} shows the frequency of number of repetitions 
among these TF-target couples.

\begin{figure}[ht]
    \centering
    \includegraphics[scale=0.5]{figures/results/TFs_count.png}
    \caption{Frequency of repeating TFs across SCENIC iterations}
    \label{fig:TFs_freq}
\end{figure}

In both figure it is clear that the variability between different iterations was high, 
since it is more frequent for TFs and TFs-target couples are less repetitive between runs. 

\begin{figure}[ht]
    \centering
    \includegraphics[scale=0.5]{figures/results/TF-gene_count.png}
    \caption{Frequency of repeating TF-target gene across SCENIC iterations}
    \label{fig:TF-target_freq}
\end{figure}

\subsection{Cellular Enrichment Score Clustering }
\subsubsection{AUCell score clustering}
Each iteration produces a different AUCell score that is clustered and visualized in a heatmap. 
Fig. \ref{fig:aucell_heatmap} is an example for such heatmap for one iteration. 
Columns are individual cell and rows are regulons. Note that regulons are named after their TF.
The $(+)$ sign attached to the regulons names symbolises activation regulatory relationship between the TF and its targets. 
The cells types annotation are attached as the top row. In this specific iteration, there were 100 predicted regulons in total. 
In fig. \ref{fig:aucell_heatmap} the cells are generally not clustered clearly by cell type, 
nonetheless some MG cells appear to cluster together, as well as some 2-5 NB cells. 
This was repetitive throughout the vast majority of iterations (the rest of the heatmaps are available in supplementary materials). 
Among the regulons that have relatively high AUCell scores in MG cells in fig.
\ref{fig:aucell_heatmap} are transcription factors from the enhancer of split gene complex $(E(spl){mbeta, m3, m5, m8}-HLH)$. 
According to FlyBase \cite{thurmond2019flybase}, these TFs are known to take part in neuronal differentiation. 
Additionally, $Myc$, $aop$ and $Mef2$ regulons were clustered with repeatedly high AUCell scores for MG cells in most iterations. 


%anterior open (\href{http://flybase.org/search/aop}{aop}) encodes a transcriptional repressor of the ETS family. 
%It acts downstream of receptor tyrosine kinase signaling to regulate cell fate transitions critical to the 
%development of many tissues including the nervous system, heart, trachea and eye. [Date last reviewed: 2019-03-07]

%Myc (\href{http://flybase.org/search/Myc}{Myc}) encodes a transcription factor that is homologous to vertebrate Myc proto-oncogenes. 
%5It contributes to cell growth, cell competition and regenerative proliferation

%Myocyte enhancer factor 2 (\href{http://flybase.org/search/Mef2}{Mef2}) encodes a protein that belongs to the 
%MADS-box family of transcription factors and is required for muscle development. It directly activates a large number of 
%muscle protein genes. It also regulates gene expression in other tissues, including the fat body and neural tissue. 
%[Date last reviewed: 2019-03-14] (\href{http://flybase.org/wiki/FlyBase:Gene_Snapshots}{FlyBase Gene Snapshot})

\begin{figure}
    \centering
    \includegraphics[width=\textwidth]{figures/results/heatmaps/aucell_run1.png}
    \caption{AUCell scores of all regulons}
    \label{fig:aucell_heatmap}
\end{figure}

\subsubsection{AUCell score clustering of most repetitive TFs}
\begin{figure}
    \centering
    \includegraphics[width=\textwidth]{figures/results/heatmaps/aucell__top80_run1.png}
    \caption{AUCell scores of core regulons}
    \label{fig:aucell_top80}
\end{figure}

% To Do: add square on the picture?
\subsubsection{AUCell score clustering with binarization}
\begin{figure}
    \centering
    \includegraphics[width=\textwidth]{figures/results/heatmaps/aucell__top80__binary_run1.png}
    \caption{Binary AUCell scores of core regulons}
    \label{fig:aucell_top80_binary}
\end{figure}

\subsubsection{AUCell score clustering grouped by cell type}
The heatmap in fig. \ref{fig:aucell_top80_grouped} was based on the same AUCell scores as in fig. \ref{fig:aucell_top80}, 
and was grouped by cell type. The value for each cell type is the average of AUCell scores for all of the cells of that same cell type.
To improve the readability of the heatmap and to enhance the visual contrasts, centering around zero and z-score scaling per row was applied. 
\begin{figure}
    \centering
    \includegraphics[width=\textwidth]{figures/results/heatmaps/mean_aucell_top80_scaled_run1.pdf}
    \caption{Average AUCell scores of core regulons for different cell types}
    \label{fig:aucell_top80_grouped}
\end{figure}
\subsection{Cytoscape GRN Visualization}
\subsubsection{Core TF-Target couples GRN}

\subsubsection{Most Repetitive TFs, with all targets GRN}
Here the thickness of the edges represents the repetitiveness of the prediction throughout SCENIC iterations. 

\section{Boolean Model Analysis}


\section{Boolean Model Extension with GRN Inferred from Single Cell Sequencing Data}